mem, tests: Add SimObject unit tests for the TreePLRU replacement policy

erin-le opened the pull request:
This PR adds SimObject unit tests for the TreePLRU replacement policy. These tests are based on the SimObject unit tests for the FIFO replacement policy in https://github.com/gem5/gem5/pull/1977, and were mainly written to see how easy or difficult it would be to write SimObject unit tests based on PR 1977.

This PR also modifies the TreePLRU replacement policy itself to avoid the following error after running a certain number of unit tests:

```bash
src/tcmalloc.cc:1460] Object was not in-use
Aborted (core dumped)
```

erin-le committed:
mem: modify member variable in TreePLRU replacement policy

This commit modifies the variable treeInstance in TreePLRU to be
a shared pointer. Without this change, SimObject unit tests for
TreePLRU will encounter an error that reads "src/tcmalloc.cc:1460]
Object was not in-use" after a certain number of tests are run.

erin-le committed:
tests, mem: Add SimObject unit tests for TreePLRU RP

This commit adds SimObject unit tests for the TreePLRU replacement
policy, based off of the SimObject unit tests in
https://github.com/gem5/gem5/pull/1977.

erin-le commented:
@odanrc: Hello, I wrote these unit tests at @BobbyRBruce's behest to see how the framework in https://github.com/gem5/gem5/pull/1977 could be used to create new unit tests for other SimObjects. Could you give this PR a review when you have time? Thank you!

odanrc commented:
@erin-le Sure, I will try to do so over the weekend. FYI, I have local commits ready with tests for LRURP, SecondChanceRP, LFURP, and MRURP - they are just waiting for the FIFORP PR to be merged.

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+
+#include <gtest/gtest.h>
+
+#include <cassert>
```
Unused

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+        params.num_leaves = num_leaves;
+        rp = std::make_shared<gem5::replacement_policy::TreePLRU>(params);
+    }
+    TreePLRUTestF(int num_entries)
```
Could you merge these 2 constructors with a default num_entries of 8?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+{
+  public:
+    std::shared_ptr<gem5::replacement_policy::TreePLRU> rp;
+    int num_leaves = 8;
```
Member variables should use camel case

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+/// Test that if there is one candidate and it is invalid, it will be
+/// the victim
+
+// I'm not sure that it makes sense to test this for TreePLRU.
```
It makes sense, because it is a behavior we expect from all RP.

We also expect the behavior ´when all entries are invalid the first candidate will always be selected, regardless of the order of the invalidations´, and there is no test for that. Could you add it (it is in the FIFO tests too)?

Finally, could you remove this comment?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+        }
+    }
+    // We use 8 entries here
+    TreePLRUVictimizationTestF() : TreePLRUTestF(), entries(num_leaves)
```
Ditto merge constructors

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+            entry.replacementData = rp->instantiateEntry();
+            candidates.push_back(&entry);
+        }
+    }
```
Strong nit: there should be space between functions to minimize the likelihood of merge conflicts

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+class TreePLRUVictimizationTestF : public TreePLRUTestF
+{
+  protected:
+    // The entries being victimized
```
Nit: use 3 slashes to make it Doxygen compliant

Although since this is a test I am not sure anyone will even see the Doxygen doc...

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    }
+};
+
+// Test only resetting one entry
```
nit: Ditto Doxygen to all tests' docs

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+// Reset entry H and entry A will be victimized.
+TEST_F(TreePLRUVictimizationTestF, GetVictimSingleResetRightmost)
+{
+    rp->reset(entries[7].replacementData);
```
Could you add here the tree state after the reset?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+// Reset entry B and entry E will be victimized.
+TEST_F(TreePLRUVictimizationTestF, GetVictimSingleResetMiddle)
+{
+    rp->reset(entries[1].replacementData);
```
Ditto tree state

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+// Test resetting no entries. The tree's nodes should all have values of 0,
+// pointing toward entry A at index 0.
+
+TEST_F(TreePLRUVictimizationTestF, GetVictimNoReset)
```
Nit: I would have put these tests before the 3 tests before, because it is simpler, and makes it easier to understand them

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+
+TEST_F(TreePLRUVictimizationTestF, GetVictimHalfReset)
+{
+    rp->reset(entries[0].replacementData);
```
nit: Could you add the tree states after each reset?

Strong nit: Could you add the tree state after the final reset?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+}
+
+// Reset all entries once, starting from the leftmost side of the tree. The
+// victimized entry should be A, as it was the least recently used.
```
Could you remove the 'as it was the least recently used' part? This is not causal. Counter example: ACDEFGHB victimizes E, despite A being the LRU.

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+
+// Test that when there is at least a single invalid entry, it will be
+// selected during the victimization
+// Taken directly from the FIFO SimObject unit tests, introduced by
```
I don't think this FIFO reference is needed. Could you remove it?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    }
+}
+
+TEST_F(TreePLRUVictimizationTestF, TestTwoTrees)
```
Could you document the goal of the test? Also, what is meant by *two* trees?

Note to myself: I need to review this test later

erin-le replied:
I might be misunderstanding the implementation of the TreePLRU replacement policy, but it looks like a new and separate PLRUTree is created each time enough entries have been created to fill up the previous tree. The "two trees" part of the test comes from the number of leaves per tree being 8, and initializing 16 entries in the test. The purpose of the test is to check that interacting with the entries of one tree doesn't affect the entries in the other tree. If you don't think this test is useful, or if there's something I'm misunderstanding, please let me know.

odanrc replied:
I see what you mean. You can mark it as resolved

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    // If the entry is correctly invalidated, the entry at index 5 will be
+    // selected.
+    // If nothing happens when invalidate is called, the entry at index 4 will
+    // be selected.
```
Nit: I would remove this comment, because we cannot 100% predict how bugged code will behave

I would also either remove the NE assertion against 4, or add this assertion against all non-5 entries.

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    ASSERT_EQ(rp->getVictim(candidates), &entries[5]);
+    ASSERT_NE(rp->getVictim(candidates), &entries[4]);
+
+    rp->reset(entries[1].replacementData);
```
What is the goal of this part of the test?

erin-le replied:
The purpose was to check that there weren't any issues with calling `reset` and `invalidate` on the same entry. If you think this isn't useful or is redundant, please let me know.

odanrc replied:
Gotcha. Could you mark it as resolved?

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    SmallTreePLRUVictimizationTestF() : TreePLRUVictimizationTestF(2) {}
+};
+
+class LargeTreePLRUVictimizationTestF : public TreePLRUVictimizationTestF
```
Nit: I am not sure this is useful. What is the goal here?

erin-le replied:
The goal was to check a larger tree size, in case there were issues that wouldn't be detected with tree sizes of 2 and 8. I can remove the test for the larger tree if you don't think it's useful.

odanrc replied:
I think that it is likely not useful (the tests with a size of 8 should be enough), but it is also not useless so you can leave it here if you want.

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    rp->invalidate(entries[1].replacementData);
+    ASSERT_EQ(rp->getVictim(candidates), &entries[1]);
+}
+
```
So far we have tested reset, invalidate and getVictim. There are no tests covering touch. Could you add them?

erin-le replied:
I added one test, `CheckTouchResetSame`, to check that `touch()` and `reset()` have the same functionality. Could you let me know if this test is sufficient, or if I should add more tests for `touch()`?

odanrc replied:
It should be good enough. You can mark this as resolved

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+}
+
+// Test the smallest tree size (2 leaves) and a large tree size (512 leaves)
+class SmallTreePLRUVictimizationTestF : public TreePLRUVictimizationTestF
```
The smallest supported tree size is 1. Could you add a small test for it, like you did for 2?

erin-le replied:
I looked into this and I'm not sure it's currently supported. I tried implementing a test and found that it failed with an exception due to the PLRUTree size being 0, as the PLRUTree is instantiated with `numLeaves - 1` as its size. Is this something that should be fixed?

odanrc replied:
I don't really think it is important to fix (what is the point of a tree PLRU with a single entry?), but we should probably create a death test to show that it is not an unknown bug, but a feature

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+    ASSERT_EQ(rp->getVictim(candidates), &entries[511]);
+}
+
+// Taken directly from the FIFO SimObject unit tests, introduced by
```
Ditto remove

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,338 @@
+{
+    gem5::ReplacementCandidates candidates;
+    ASSERT_DEATH(rp->getVictim(candidates), "");
+}
```
Add death test for num leaves <= 0.

erin-le committed:
mem, tests: address some review comments for TreePLRU unit tests

erin-le committed:
mem, tests: add test to check touch() for TreePLRU

touch() and reset() have the same functionality for the TreePLRU
replacement policy. This commit adds a unit test to check that
touch() has the same functionality as reset().

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,402 @@
+
+/// Test that when all entries are invalid the first candidate will always be
+/// selected, regardless of the order of the invalidations. The order of
+/// invalidations of this test is reversed in comparison to the other test.
```
Which other test? I think this is an artifact from a copy paste

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,402 @@
+    ///  __0__   __0__
+    /// _0_ _0_ _0_ _0_
+    /// A B C D E F G H
+    ASSERT_EQ(rp->getVictim(candidates), expected_victim);
```
This line is a duplicate of test GetVictimNoReset, so I suggest reordering these 2 tests and removing this line

odanrc commented on src/mem/cache/replacement_policies/tree_plru_rp.test.cc:
```diff
@@ -0,0 +1,402 @@
+    ASSERT_EQ(rp->getVictim(candidates), &entries[7]);
+}
+
+/// In the definition of the TreePLRU replacement policy, `reset()` has the
```
nit: In tests one shouldn't refer to code implementations, because code changes; one should write tests based on the expected behavior. For example, here someone could change the code so that touch calls reset instead of the other way around, and this comment would become outdated. I would simply say that touching and resetting should behave the same for this RP.

erin-le committed:
tests: Make additional changes to TreePLRU unit tests

This commit adds a death test for a TreePLRU replacement policy
that only accommodates one leaf. It also modifies some comments
and makes small changes.

giactra commented:
Hi all, thanks for the test! I just wanted to provide this feedback. I believe I once encountered an error in the TrePLRU if the associativity of the cache was set to 1 (direct mapped). Would it be possible to add this test as well?

erin-le commented:
@giactra: Sorry for the delayed response. At one point, I did try to add a (normal) test for num_leaves = 1, but ran into some issues with it (see [this comment chain](https://github.com/gem5/gem5/pull/2548#discussion_r2295847454)) and added a death test instead (see [here](https://github.com/gem5/gem5/pull/2548/files#diff-c25e502afe6ff2d85469985ccf115c215edef02188666ed47f1480733425b9dcR408), at the bottom of the file).

erin-le committed:
Merge branch 'develop' into TreePLRU-simobject-unit-tests
